### PR TITLE
Amp-bind: Cleanup removed bindings from bind evaluator

### PR DESCRIPTION
**Pull request by kmh287**
The bind evaluator doesn't currently delete expressions from its cache even when the expressions are no longer used for any bind attributes on the page. This PR clears the cache when bindings are removed. The caller should only call with expression strings that are no longer used to bind any attributes on the page. bind-impl already does this [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-bind/0.1/bind-impl.js#L275)

Detecting bindings in amp-live-lists introduces the potential for this cache to grow without bound.

/to @choumx

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 1]
```diff
@@ -81,7 +81,10 @@ export class BindEvaluator {
   removeBindingsWithExpressionStrings(expressionStrings) {
     const expressionsToRemove = Object.create(null);
     for (let i = 0; i < expressionStrings.length; i++) {
-      expressionsToRemove[expressionStrings[i]] = true;
+      const expressionString = expressionStrings[i];
+      delete this.expressions_[expressionString];
+      // Use a map instead of an array to make filtering bindings_ faster
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Nit: This comment is unnecessary.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 2]
```diff
@@ -81,7 +81,10 @@ export class BindEvaluator {
   removeBindingsWithExpressionStrings(expressionStrings) {
     const expressionsToRemove = Object.create(null);
     for (let i = 0; i < expressionStrings.length; i++) {
-      expressionsToRemove[expressionStrings[i]] = true;
+      const expressionString = expressionStrings[i];
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Nit: `forEach` would be slightly more concise (for some negligible perf cost).

Comment (kmh287) [inline extensions/amp-bind/0.1/bind-evaluator.js]: sure

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 3]
```diff
@@ -64,12 +69,24 @@ describe('BindEvaluator', () => {
...
-    const bindingDef = {
+  it('should NOT cache expressions that have been removed', () => {
+    expect(numberOfCachedExpressions()).to.equal(0);
+    evaluator.addBindings([{
       tagName: 'P',
       property: 'text',
       expressionString: 'oneplusone + 2',
-    };
+    }, {
+      tagNme: 'A',
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: tagName

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 4]
```diff
@@ -64,12 +69,24 @@ describe('BindEvaluator', () => {
...
+      property: 'href',
+      expressionString: 'url',
+    }]);
+    // Three bound elements
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Remove inaccurate comment.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 5]
```diff
@@ -64,12 +69,24 @@ describe('BindEvaluator', () => {
     expect(numberOfBindings()).to.equal(0);
   });
 
-  it('should evaluate expressions given a scope with needed bindings', () => {
-    const bindingDef = {
+  it('should NOT cache expressions that have been removed', () => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Nit: should clean up removed expressions from cache

**Pull request merged**